### ISSUE TITLE
Setup CI for v2

### ISSUE THREAD

**Issue by jpkrohling** [jaegertracing/jaeger#2883]
## Requirement - what kind of business use case are you trying to solve?

We need to set up a custom CI infra for the v2, as there are things from v1 that we might not need, do differently, or that don't exist yet.

## Proposal - what do you suggest to solve the problem or improve the existing situation?

- [ ] Set the current CI to ignore PRs that include only changes to v2
- [ ] Create a new Makefile inside the v2 with the targets we need for v2
- [ ] Create a new v2-specific GitHub workflows

**Comment by yurishkuro** [comment]
It sounds like we are trying the "boil the ocean" approach. Are we saying that we're not going to reuse any of the existing code? I hope not. But if we only do the refactoring to support both v1 and v2 then there will be changes that must be tested for both versions.

**Comment by jpkrohling** [comment]
We are definitely going to reuse all of the "business'' code, but there are things that we need only for v2, and things we need only for v1, hence the split. Common things can be placed at the main workflow dir, and other things are so simple that we can simply duplicate without fear. My current idea is to get something building as part of v2 alone, and then later iterate and see what we can/should reuse. Small, continuous steps, so that things at least move a bit :-)

**Comment by yurishkuro** [comment]
Any duplication is undesirable imo. Even the in-memory exporter needs to share code with v1, so it's unavoidable problem that we need to solve.

**Comment by jpkrohling** [comment]
>Any duplication is undesirable imo.

Why is that? There are fundamental disparities between the otelcol code base and Jaeger, like in the handling of their own metrics. To keep current consumers of the v1 code working, I would simply copy the v1 code to v2, and change it there to fit the new code base. Whenever possible, we can share code, but making v1 a dependency of v2 is only going to give us headaches in the future.

**Comment by yurishkuro** [comment]
Copying code means we're going to have two copies to maintain, fix bugs in two places, duplicate new features being added, over several months at least. Not to mention that if prod-quality v1 is copied and modified, the confidence in the copy being prod-quality falls dramatically.

Pavol was able to implement the first iteration of OTEL binaries without duplicating code, with some refactoring. It's always better to refactor code to accommodate the migration than to copy it. Things like metrics usually can be done as decorators, and when they must be mixed into biz code they can be abstracted by an interface.

**Comment by jpkrohling** [comment]
Agree that we'd have to fix bugs in two places, but we are not going to implement features for v1 only because we are adding a feature for v2. 

>Not to mention that if prod-quality v1 is copied and modified, the confidence in the copy being prod-quality falls dramatically.

Not sure I get this part. Care to expand?

>Pavol was able to implement the first iteration of OTEL binaries without duplicating code, with some refactoring.

Indeed, but we are not only having OpenTelemetry binaries, we are also having a major version change. We just can't have v2 depend on v1, it doesn't make any sense :-) We decided to go for a v2 also because, despite Pavol's heroic work, we ended up with dependencies between the projects that weren't easy to understand, or even manageable in the long term (see the Thrift dependencies problem that eventually led to the OpenTelemetry code to be removed from this repo. A v2 that copies code from v1 will allow us to clearly define the boundaries and dependencies, without having to take backward compatibility into consideration, at least for the API.

**Comment by yurishkuro** [comment]
> Not sure I get this part. Care to expand?

A code that is used in production - hi confidence. The same code that's copied & modified & not used in production - low confidence.

I am not getting your arguments in favor of copying. The situation with dependencies was due to a circular dependency where OTEL directly imports Jaeger packages for some receivers while we also depend on OTEL for framework. I am still not clear how we plan to address that circular dependency - it's not solved by code copying in any way.

And we're not just building a new version, we're building a backwards-compatible version, so all server endpoints and all storage implementations have to work the same way (which is the majority of the code), the only thing that is changing is the configuration and wiring up of the existing components.

**Comment by jpkrohling** [comment]
>A code that is used in production - hi confidence. The same code that's copied & modified & not used in production - low confidence.

Makes sense, thanks.

>I am not getting your arguments in favor of copying.

Removing otel a bit from the picture, what I'm saying is that it might be easier to evolve the code base if v2 is independent of v1. You might be right, it might be possible to have v2 depend on v1 and get away with it, and I'm willing to give it a try.

>I am still not clear how we plan to address that circular dependency - it's not solved by code copying in any way.

The current idea is to remove Jaeger from the open-telemetry/opentelemetry-collector repository while keeping it in the main distribution. What I'm suggesting to them is that the Jaeger code is moved to this repository here.

>And we're not just building a new version, we're building a backwards-compatible version, so all server endpoints and all storage implementations have to work the same way (which is the majority of the code), the only thing that is changing is the configuration and wiring up of the existing components.

Right, we are keeping backward compatibility when it comes to our external runtime interfaces, both at the receiving end and on the storage side. We have no obligation though to keep API compatibility for code consuming this repository as a library, which was my point above. Health check and bounded queue are two examples that come to my mind.

In summary, I think we can agree that we'll try to use as much as we can from the current code base, without moving code.

I would still like to have a separate set of CI scripts, especially because how we build the v2 is going to be significantly different from v1: there's no main code on v2, the binaries are coming based on the manifest files, assembled by the opentelemetry-collector-builder. I would also take this opportunity to make the release process simpler, by making use of goreleaser.

**Comment by yurishkuro** [comment]
I agree about separate CI scripts, since we're generating a whole set of binaries that need to be tested.

> The current idea is to remove Jaeger from the open-telemetry/opentelemetry-collector repository while keeping it in the main distribution.

I thought it was never there, they were importing directly from our repo.

**Comment by jpkrohling** [comment]
That's part of the confusion: they have OpenTelemetry components (receivers, exporters) that consume our code. The proposal is to move the actual components to our repo as well, along with the "translators". Concretely, these:

https://github.com/open-telemetry/opentelemetry-collector/tree/main/translator/trace/jaeger
https://github.com/open-telemetry/opentelemetry-collector/tree/main/receiver/jaegerreceiver
https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/jaegerexporter

**Comment by yurishkuro** [comment]
I see, that makes sense. But if that move happens, are they still going to import these components statically, or only using the builder config? The latter would avoid direct go.mod circular dependency.

**Comment by jpkrohling** [comment]
I think they are building the binaries with something similar to the builder, but I think nobody is 100% sure at this point.

**Comment by jpkrohling** [comment]
I just sent a PR for this issue, taking into consideration what we have discussed here. I still have a bad feeling about having shared code between v1 and v2 and thought of some cases where it'll be tricky to manage. My main reference about Go modules for v2 is this: https://blog.golang.org/v2-go-modules, which does recommend copying the whole codebase to a new v2 directory.

As discussed in this issue here, the current agreement is to have some code in the current place, like the [memory storage plugin](https://github.com/jaegertracing/jaeger/blob/master/plugin/storage/memory/memory.go), and the new code under v2, including the otelcol exporters and receivers. For a change in that file to be consumed by v2 (storage exporter, for instance), we'll then need:

1. perform the desired code changes to the storage plugin
1. to test it, use a `replace` directive in the v2/go.mod file, pointing to the v1 (root) directory
1. once its done, send a PR for the storage plugin
1. once that PR is merged, the v2/go.mod needs to be updated to consume the changes
1. change the consumer code, and open a PR, including the v2/go.mod changes

To release a version with the desired change, we need to:

1. release a new v1 version
1. change the v2/go.mod to use the new v1 release in the dependency
1. release a new v2 version

This is all (probably) doable, and I understand the code duplication argument, but I wonder how maintainable this whole setup is going to be in the future when the only real advantage is keeping one version of the source file to be shared by two different consumers, with potentially conflicting future requirements.

To be honest, a whole new branch (v2) is what makes the most sense to me, as we can tag a commit from this new branch and be sure that it contains only code related to v2. Similarly, we can be sure that a tag for v1 only has v1 code in it. But between this personal preference and following the advice of Go module developers, I'd probably stay with the experts, unless more people agree with me.
